google/cloud/firestore_bundle/bundle.py

```diff
@@ -0,0 +1,439 @@
+
+        Returns:
+            FirestoreBundle: self
+
+        """
+        if isinstance(document_or_query_name, DocumentSnapshot):
+            assert query is None
+            self.add_document(document_or_query_name)
+        elif isinstance(document_or_query_name, str):
+            assert query is not None
+            self.add_named_query(document_or_query_name, query)
+        else:
+            raise ValueError(
+                "Bundle.add accepts either a standalone DocumentSnapshot or "
+                "a name (string) and a Query."
+            )
+        return self
+
+    def add_document(
+        self, snapshot: DocumentSnapshot, query_name: Optional[str] = None,
+    ) -> "FirestoreBundle":
+        """Adds a document to the bundle.
+
+        Args:
+            snapshot (DocumentSnapshot): The fully-loaded Firestore document to
+                be preserved.
+            query_name (Optional[str]): If provided, also establishes a link
+                between the provided document and the referenced query.
+
+        Example:
+
+            from google.cloud import firestore
+
+            db = firestore.Client()
+            collection_ref = db.collection(u'users')
+
+            bundle = firestore.FirestoreBundle('my bundle')
+            bundle.add_document(collection_ref.documents('some_id').get())
+
+        Returns:
+            FirestoreBundle: self
+        """
+        original_document: Optional[_BundledDocument]
+        original_queries: Optional[List[str]] = []
+        full_document_path: str = snapshot.reference._document_path
+
+        original_document = self.documents.get(full_document_path)
+        if original_document:
+            original_queries = original_document.metadata.queries  # type: ignore
+
+        should_use_snaphot: bool = (
+            original_document is None
+            # equivalent to:
+            #   `if snapshot.read_time > original_document.snapshot.read_time`
+            or _helpers.compare_timestamps(
+                snapshot.read_time, original_document.snapshot.read_time,
+            )
+            >= 0
+        )
+
+        if should_use_snaphot:
+            self.documents[full_document_path] = _BundledDocument(
+                snapshot=snapshot,
+                metadata=BundledDocumentMetadata(
+                    name=full_document_path,
+                    read_time=snapshot.read_time,
+                    exists=snapshot.exists,
+                    queries=original_queries,
+                ),
+            )
+
+        if query_name:
+            bundled_document = self.documents.get(full_document_path)
+            bundled_document.metadata.queries.append(query_name)  # type: ignore
+
+        self._update_last_read_time(snapshot.read_time)
+        # Flush this if it was cached
+        self._deserialized_metadata = None
+        return self
+
+    def add_named_query(self, name: str, query: BaseQuery) -> "FirestoreBundle":
```

Comment:
Other SDKs take QuerySnapshot instead of Query. We made this choice to avoid having to surface errors properly when building the bundle, if the query run to error.

If I read this correctly, the query here will be run by the bundle builder? Please use snapshot instead, for above reason and for consistency.

Reply:
Sadly, the Python SDK has no concept of a `QuerySnapshot` 😭 . I very much wanted to use that, instead.

Reply:
Sorry, I have no idea `QuerySnapshot` is not a thing in Python...I am surprised.

However, we could potentially accept whatever `Query.get` returns here (looks like it returns a list of `DocumentSnapshot`)?

The benefit, as stated in the original comment, is you do not need to worry about reporting error when `build()`. Otherwise, we should try to tell the users exactly which named query failed to fetch when we `build()`, we also need to decide if one of the query failed, should `build()` still proceed.